src/main/scala/is/hail/distributedmatrix/BlockMatrixIsDistributedMatrix.scala

```diff
@@ -334,38 +289,151 @@ object BlockMatrixIsDistributedMatrix extends DistributedMatrix[BlockMatrix] {
+    }
+      .fold(0L)(_ + _)
+
+    info(s"wrote $blockCount blocks") // should equal number of partitions
+  }
 
   /**
-    * Reads a BlockMatrix matrix written by {@code write} at location {@code
-    * uri}.
+    * Reads a BlockMatrix matrix written by {@code write} at location
+    * {@code uri}.
     *
     **/
   def read(hc: HailContext, uri: String): M = {
     val hadoop = hc.hadoopConf
     hadoop.mkDir(uri)
-
-    val rdd = hc.sc.sequenceFile[PairWriter, MatrixWriter](uri+matrixRelativePath).map { case (pw, mw) =>
-      ((pw.i, pw.j), mw.toDenseMatrix(): Matrix)
-    }
-
+    
     val BlockMatrixMetadata(rowsPerBlock, colsPerBlock, numRows, numCols) =
-      hadoop.readTextFile(uri+metadataRelativePath) { isr  =>
+      hadoop.readTextFile(uri + metadataRelativePath) { isr  =>
         jackson.Serialization.read[BlockMatrixMetadata](isr)
       }
 
-    new BlockMatrix(rdd, rowsPerBlock, colsPerBlock, numRows, numCols)
+    val nRowBlocks = ((numRows - 1) / rowsPerBlock).toInt + 1
+    val nColBlocks = ((numCols - 1) / colsPerBlock).toInt + 1
+    
+    val blocks = new ReadBlocksRDD(hc.sc, uri, nRowBlocks, nColBlocks)
+    
+    new BlockMatrix(blocks, rowsPerBlock, colsPerBlock, numRows, numCols)
   }
 }
 
+case class ReadBlocksRDDPartition(index: Int) extends Partition
+
+class ReadBlocksRDD(sc: SparkContext, uri: String, nRowBlocks: Int, nColBlocks: Int) extends RDD[((Int, Int), Matrix)](sc, Nil) {
+  private val nBlocks= nRowBlocks * nColBlocks
+  assert(nBlocks >= nRowBlocks && nBlocks >= nColBlocks)
+  
+  override def getPartitions: Array[Partition] =
+    Array.tabulate(nBlocks)(i => ReadBlocksRDDPartition(i))
+  
+  private val sHadoopConfBc = sc.broadcast(new SerializableHadoopConfiguration(sc.hadoopConfiguration))
+
+  override def compute(split: Partition, context: TaskContext): Iterator[((Int, Int), Matrix)] = {
+    val d = digitsNeeded(nBlocks)
+    val i = split.index
+    assert(i >= 0 && i < nBlocks)
+
+    val is = i.toString
+    assert(is.length <= d)
+    val pis = StringUtils.leftPad(is, d, "0")
+
+    Iterator.single(sHadoopConfBc.value.value.readDataFile(uri + "/blocks/block-" + pis) { in =>
+      val mw = new MatrixWriter()
+      mw.readFields(in)
+      
+      // block indices are column major
+      ((i % nRowBlocks, i / nRowBlocks), mw.toDenseMatrix())
+    })
+  }
+
+  override val partitioner: Option[Partitioner] = Some(GridPartitioner(nRowBlocks, nColBlocks))
+}
+
 // must be top-level for Jackson to serialize correctly
 case class BlockMatrixMetadata(rowsPerBlock: Int, colsPerBlock: Int, numRows: Long, numCols: Long)
+
+class PairWriter(var i: Int, var j: Int) extends Writable {
+  def this() {
+    this(0,0)
+  }
+
+  def write(out: DataOutput) {
+    out.writeInt(i)
+    out.writeInt(j)
+  }
+
+  def readFields(in: DataInput) {
+    i = in.readInt()
+    j = in.readInt()
+  }
+}
+
+class MatrixWriter(var rows: Int, var cols: Int, var m: Array[Double]) extends Writable {
```

Comment:
Meh.  Not that it really matters, but I wouldn't allocate this class.  Just make readDenseMatrix/writeDenseMatrix routines (e.g. on Rich{Input, Output}Stream) or just routines on an object somewhere.  This object mutation pattern comes from Hadoop and there's no reason to keep it here.